File: codeforces/clothes.cc
#include <iostream>
#include <limits>

using namespace std;
using ll = long long;

int prices[100];
bool match[100][100];

int main() {
  int n, m, a, b;
  cin >> n >> m;
  ll MAXVAL = numeric_limits<ll>::max(), minPrice = MAXVAL;

  for (int i = 0; i < n; i++)
    cin >> prices[i];
  for (int i = 0; i < m; i++) {
    cin >> a >> b;
    match[a - 1][b - 1] = match[b - 1][a - 1] = true;
  }
  for (int i = 0; i < n; i++) {
    for (int j = i + 1; j < n; j++) {
      for (int k = j + 1; k < n; k++) {
        if (match[i][j] && match[j][k] && match[k][i])
          minPrice = min(minPrice, ll(prices[i] + prices[j] + prices[k]));
      }
    }
  }

  cout << (minPrice == MAXVAL ? -1 : minPrice);
  return 0;
}

File: codeforces/clothes.py
from collections import defaultdict
from itertools import combinations

# visualized this as a graph problem so gave it a shot
# it worked well but there is more room for improvement in the main for loop

MAXVAL = 10000000
n, m = map(int, input().split())
price = list(map(int, input().split()))
match = defaultdict(list)
minPrice = MAXVAL

for _ in range(m):
    i, j = map(int, input().split())
    match[i].append(j)
    match[j].append(i)

match = dict(sorted(match.items(), key=lambda x: len(x[1]), reverse=True))

for item in match.items():
    if len(item[1]) < 2:
        break
    i = item[0]
    for temp in combinations(item[1], 2):
        j, k = temp
        if k not in match[j]:
            continue
        minPrice = min(minPrice, price[i-1] + price[j-1] + price[k-1])

if minPrice == MAXVAL:
    print(-1)
else:
    print(minPrice)
